File: packages-gen/dart/lang/lib/common-trace.dart
import 'package:xtalk_lang/common-data.dart' as xtd;

trace_log() {
  if((() {
    var dart_truthy__39899 = !(null == __globals__["TRACE"]);
    return (null != dart_truthy__39899) && (false != dart_truthy__39899);
  })()){
    return __globals__["TRACE"];
  }
  else{
    __globals__["TRACE"] = <dynamic>[];
    return __globals__["TRACE"];
  }
}

trace_log_clear() {
  __globals__["TRACE"] = <dynamic>[];
  return __globals__["TRACE"];
}

trace_log_add(data, tag, opts) {
  var log = trace_log();
  var m = xtd.obj_assign(<dynamic, dynamic>{
    "tag":tag,
    "data":data,
    "time":DateTime.now().millisecondsSinceEpoch
  },opts);
  log.add(m);
  return log.length;
}

trace_filter(tag) {
  return xtd.arr_filter(trace_log(),(e) {
    return tag == e["tag"];
  });
}

trace_last_entry(tag) {
  var log = trace_log();
  if(null == tag){
    return log[log.length + -1];
  }
  else{
    var tagged = trace_filter(tag);
    return tagged[tagged.length + -1];
  }
}

trace_data(tag) {
  return xtd.arr_map(trace_log(),(e) {
    return e["data"];
  });
}

trace_last(tag) {
  return (trace_last_entry(tag))["data"];
}

trace_run(f) {
  trace_log_clear();
  f();
  return trace_log();
}

File: packages-gen/js/libs/xt-lang/common-trace.js
const xtd = require("@xtalk/lang/common-data.js")

function trace_log(){
  if(!(null == globalThis["TRACE"])){
    return globalThis["TRACE"];
  }
  else{
    globalThis["TRACE"] = [];
    return globalThis["TRACE"];
  }
}

function trace_log_clear(){
  globalThis["TRACE"] = [];
  return globalThis["TRACE"];
}

function trace_log_add(data,tag,opts){
  let log = trace_log();
  let m = xtd.obj_assign({"tag":tag,"data":data,"time":Date.now()},opts);
  log.push(m);
  return log.length;
}

function trace_filter(tag){
  return trace_log().filter(function (e){
    return tag == e["tag"];
  });
}

function trace_last_entry(tag){
  let log = trace_log();
  if(null == tag){
    return log[log.length + -1];
  }
  else{
    let tagged = trace_filter(tag);
    return tagged[tagged.length + -1];
  }
}

function trace_data(tag){
  return trace_log().map(function (e){
    return e["data"];
  });
}

function trace_last(tag){
  return (trace_last_entry(tag))["data"];
}

function trace_run(f){
  trace_log_clear();
  f();
  return trace_log();
}

module.exports = {
  ["trace_log"]:trace_log,
  ["trace_log_clear"]:trace_log_clear,
  ["trace_log_add"]:trace_log_add,
  ["trace_filter"]:trace_filter,
  ["trace_last_entry"]:trace_last_entry,
  ["trace_data"]:trace_data,
  ["trace_last"]:trace_last,
  ["trace_run"]:trace_run
}